File: credit/credit.py
from cs50 import get_string


def get_input(prompt):
    # Function to prompt user for card number

    while True:
        card_number = get_string(prompt)
        if card_number.isdecimal():
            return card_number


def validate_card_number(number):
    # Function to validate card number entered by user

    sum = 0

    # Determine sum of alternate digits starting from last digit
    for i in range(len(number) - 1, -1, -2):
        sum += int(number[i])

    # Determine sum of double of alternate digits starting from penultimate digit
    for i in range(len(number) - 2, -1, -2):
        double = int(number[i]) * 2

        # Add 2 digits of double if it has more than 1 digit
        if len(str(double)) == 2:
            double = int(str(double)[0]) + int(str(double)[1])

        sum += double

    # Perform further checks if sum is divisible by 10, else return false
    if not sum % 10:
        length = len(number)

        if length == 13:
            return "VISA" if number[0:1] == "4" else "INVALID"
        elif length == 15:
            return "AMEX" if number[0:2] == "34" or number[0:2] == "37" else "INVALID"
        elif length == 16:
            n = int(number[0:2])
            if n >= 51 and n <= 55:
                return "MASTERCARD"
            elif n >= 40 and n <= 49:
                return "VISA"
            else:
                return "INVALID"
        else:
            return "INVALID"

    else:
        return "INVALID"


def main():
    card_number = get_input("Number: ")
    print(validate_card_number(card_number))


if __name__ == "__main__":
    # Call main function
    main()

File: credit/credit.c
// Validates credit card numbers

#include <cs50.h>
#include <stdio.h>

long get_input(void);
string validate_number(long);

int main(void)
{
    long number = get_input();
    printf("%s\n", validate_number(number));
}

// Prompts the user for a positive card number
long get_input(void)
{
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0);
    return number;
}

// Validates the card number
string validate_number(long number)
{
    int digits = 0, sum = 0;
    int temp;
    while (number > 0)
    {
        // Takes a digit or its double based on its position
        int digit = (digits == 0 || digits % 2 == 0) ? number % 10 : number % 10 * 2;
        // Adds the two digits if the result > 9, i.e. 12 would be 1 + 2 = 3
        if (digit > 9)
        {
            digit = 1 + digit % 10;
        }
        sum += digit;
        // Calculates the total number of digits in the card number
        digits++;
        // Stores the first 2 digits of the card number for later validation
        if (number > 9 && number < 100)
        {
            temp = number;
        }
        // Removes the last digit of the card number before the next iteration
        number /= 10;
    }
    if (sum % 10 == 0)
    {
        switch (digits)
        {
            case 13:
                return temp >= 40 && temp <= 49 ? "VISA" : "INVALID";
            case 15:
                return temp == 34 || temp == 37 ? "AMEX" : "INVALID";
            case 16:
                if (temp >= 51 && temp <= 55)
                {
                    return "MASTERCARD";
                }
                else if (temp >= 40 && temp <= 49)
                {
                    return "VISA";
                }
                else
                {
                    return "INVALID";
                }
            default:
                return "INVALID";
        }
    }
    else
    {
        return "INVALID";
    }
}
